File: android/src/main/kotlin/dev/asdevs/signalr_flutter/SignalR.kt
package dev.asdevs.signalr_flutter

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.MethodChannel.Result
import microsoft.aspnet.signalr.client.*
import microsoft.aspnet.signalr.client.hubs.HubConnection
import microsoft.aspnet.signalr.client.hubs.HubProxy
import microsoft.aspnet.signalr.client.transport.LongPollingTransport
import microsoft.aspnet.signalr.client.transport.ServerSentEventsTransport

enum class CallMethod(val value: String) {
    ConnectToServer("connectToServer"),
    Reconnect("reconnect"),
    Stop("stop"),
    IsConnected("isConnected"),
    ListenToHubMethod("listenToHubMethod"),
    InvokeServerMethod("invokeServerMethod")
}

object SignalR {
    private lateinit var connection: HubConnection
    private lateinit var hub: HubProxy

    fun connectToServer(url: String, hubName: String, queryString: String, headers: Map<String, String>, transport: Int, hubMethods: List<String>, result: Result) {
        try {
            connection = if (queryString.isEmpty()) {
                HubConnection(url)
            } else {
                HubConnection(url, queryString, true, Logger { _: String, _: LogLevel ->
                })
            }

            if (headers.isNotEmpty()) {
                val cred = Credentials() { request ->
                    request.headers = headers
                }
                connection.credentials = cred
            }
            hub = connection.createHubProxy(hubName)

            hubMethods.forEach { methodName ->
                hub.on(methodName, { res ->
                    Handler(Looper.getMainLooper()).post {
                        SignalRFlutterPlugin.channel.invokeMethod("NewMessage", listOf(methodName, res))
                    }
                }, Any::class.java)
            }

            connection.connected {
                Handler(Looper.getMainLooper()).post {
                    SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", listOf(connection.state.name, connection.connectionId, null))
                }
            }

            connection.reconnected {
                Handler(Looper.getMainLooper()).post {
                    SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", listOf(connection.state.name, connection.connectionId, null))
                }
            }

            connection.reconnecting {
                Handler(Looper.getMainLooper()).post {
                    SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", listOf(connection.state.name, connection.connectionId, null))
                }
            }

            connection.closed {
                Handler(Looper.getMainLooper()).post {
                    SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", listOf(connection.state.name, null, null))
                }
            }

            connection.connectionSlow {
                Handler(Looper.getMainLooper()).post {
                    SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", listOf("ConnectionSlow", connection.connectionId, null))
                }
            }

            connection.error { handler ->
                Handler(Looper.getMainLooper()).post {
                    SignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", listOf("ConnectionError", null, handler.localizedMessage))
                }
            }

            when (transport) {
                1 -> connection.start(ServerSentEventsTransport(connection.logger))
                2 -> connection.start(LongPollingTransport(connection.logger))
                else -> {
                    connection.start()
                }
            }

            result.success(connection.connectionId)
        } catch (ex: Exception) {
            result.error("Error", ex.localizedMessage, null)
        }
    }

    fun reconnect(result: Result) {
        try {
            connection.start()
            result.success(connection.connectionId)
        } catch (ex: Exception) {
            result.error(ex.localizedMessage, ex.stackTrace.toString(), null)
        }
    }

    fun stop(result: Result) {
        try {
            connection.stop()
        } catch (ex: Exception) {
            result.error(ex.localizedMessage, ex.stackTrace.toString(), null)
        }
    }

    fun isConnected(result: Result) {
        try {
            if (this::connection.isInitialized) {
                when (connection.state) {
                    ConnectionState.Connected -> result.success(true)
                    else -> result.success(false)
                }
            } else {
                result.success(false)
            }
        } catch (ex: Exception) {
            result.error("Error", ex.localizedMessage, null)
        }
    }

    fun listenToHubMethod(methodName: String, result: Result) {
        try {
            hub.on(methodName, { res ->
                Handler(Looper.getMainLooper()).post {
                    SignalRFlutterPlugin.channel.invokeMethod("NewMessage", listOf(methodName, res))
                }
            }, Any::class.java)
        } catch (ex: Exception) {
            result.error("Error", ex.localizedMessage, null)
        }
    }

    fun invokeServerMethod(methodName: String, args: List<Any>, result: Result) {
        try {
            val res: SignalRFuture<Any> = hub.invoke(Any::class.java, methodName, *args.toTypedArray())

            res.done { msg: Any? ->
                Handler(Looper.getMainLooper()).post {
                    result.success(msg)
                }
            }

            res.onError { throwable ->
                Handler(Looper.getMainLooper()).post {
                    result.error("Error", throwable.localizedMessage, null)
                }
            }
        } catch (ex: Exception) {
            result.error("Error", ex.localizedMessage, null)
        }
    }
}

File: ios/Classes/SignalR.swift
//
//  SignalR.swift
//  signalR
//
//  Created by Ayon Das on 23/07/20.
//

import Foundation

enum CallMethod : String {
  case connectToServer, reconnect, stop, invokeServerMethod, listenToHubMethod
}

class SignalRWrapper {

  static let instance = SignalRWrapper()
  private var hub: Hub!
  private var connection: SignalR!

  func connectToServer(baseUrl: String, hubName: String, transport: Int, queryString : String, headers: [String: String], hubMethods: [String], result: @escaping FlutterResult) {
    connection = SignalR(baseUrl)

    if !queryString.isEmpty {
      let qs = queryString.components(separatedBy: "=")
      connection.queryString = [qs[0]:qs[1]]
    }

    if transport == 1 {
      connection.transport = Transport.serverSentEvents
    } else if transport == 2 {
      connection.transport = Transport.longPolling
    }

    if headers.count > 0 {
      connection.headers = headers
    }

    hub = connection.createHubProxy(hubName)

    hubMethods.forEach { (methodName) in
      hub.on(methodName) { (args) in
        SwiftSignalRFlutterPlugin.channel.invokeMethod("NewMessage", arguments: [methodName, args?[0]])
      }
    }

    connection.starting = { [weak self] in
      print("SignalR Connecting. Current Status: \(String(describing: self?.connection.state.stringValue))")
      SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: self?.connection.state.stringValue)
    }

    connection.reconnecting = { [weak self] in
      print("SignalR Reconnecting. Current Status: \(String(describing: self?.connection.state.stringValue))")
      SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: self?.connection.state.stringValue)
    }

    connection.connected = { [weak self] in
      print("SignalR Connected. Connection ID: \(String(describing: self?.connection.connectionID))")
      SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: self?.connection.state.stringValue)
    }

    connection.reconnected = { [weak self] in
      print("SignalR Reconnected...")
      print("Connection ID: \(String(describing: self?.connection.connectionID))")
      SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: self?.connection.state.stringValue)
    }

    connection.disconnected = { [weak self] in
      print("SignalR Disconnected...")
      SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: self?.connection.state.stringValue)
    }

    connection.connectionSlow = {
      print("Connection slow...")
      SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: "Slow")
    }

    connection.error = { [weak self] error in
      print("Error: \(String(describing: error))")
      SwiftSignalRFlutterPlugin.channel.invokeMethod("ConnectionStatus", arguments: self?.connection.state.stringValue)

      if let source = error?["source"] as? String, source == "TimeoutException" {
        print("Connection timed out. Restarting...")
        self?.connection.start()
      }
    }

    connection.start()
    result(true)
  }

  func reconnect(result: @escaping FlutterResult) {
    if let connection = self.connection {
      connection.connect()
    } else {
      result(FlutterError(code: "Error", message: "SignalR Connection not found or null", details: "Start SignalR connection first"))
    }
  }

  func stop(result: @escaping FlutterResult) {
    if let connection = self.connection {
      connection.stop()
    } else {
      result(FlutterError(code: "Error", message: "SignalR Connection not found or null", details: "Start SignalR connection first"))
    }
  }

  func listenToHubMethod(methodName : String, result: @escaping FlutterResult) {
    if let hub = self.hub {
      hub.on(methodName) { (args) in
        SwiftSignalRFlutterPlugin.channel.invokeMethod("NewMessage", arguments: [methodName, args?[0]])
      }
    } else {
      result(FlutterError(code: "Error", message: "SignalR Connection not found or null", details: "Connect SignalR before listening a Hub method"))
    }
  }

  func invokeServerMethod(methodName: String, arguments: [Any]? = nil, result: @escaping FlutterResult) {
    do {
      if let hub = self.hub {
        try hub.invoke(methodName, arguments: arguments, callback: { (res, error) in
          if let error = error {
            result(FlutterError(code: "Error", message: String(describing: error), details: nil))
          } else {
            result(res)
          }
        })
      } else {
        throw NSError.init(domain: "NullPointerException", code: 0, userInfo: [NSLocalizedDescriptionKey : "Hub is null. Initiate a connection first."])
      }
    } catch {
      result(FlutterError.init(code: "Error", message: error.localizedDescription, details: nil))
    }
  }
}

